File: cpp/22_Generate_Parentheses.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void genPs(int open, int close, vector<string>& ret, string& s, int i) {
        if (open < 0 || close < 0 || close < open) return;

        if (open == 0 && close == 0) {
            ret.push_back(s);
            return;
        }

        s[i] = '(';
        genPs(open - 1, close, ret, s, i + 1);

        s[i] = ')';
        genPs(open, close - 1, ret, s, i + 1);
    }

    vector<string> generateParenthesis(int n) {
        vector<string> ret;
        string s(n << 1, ' ');
        genPs(n, n, ret, s, 0);
        return ret;
    }
};

File: cpp/22_Generate_Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        s = [''] * (2 * n)

        def gen(open, close, i):
            if open < 0 or close < 0 or close < open:
                return

            if open == 0 and close == 0:
                res.append(''.join(s))
                return

            s[i] = '('
            gen(open - 1, close, i + 1)

            s[i] = ')'
            gen(open, close - 1, i + 1)

        gen(n, n, 0)
        return res
